packages/cozy-client/src/Schema.js

```diff
@@ -52,13 +59,43 @@ const normalizeDoctypeSchema = doctypeSchema => {
  */
 class Schema {
   constructor(schemaDefinition = {}, client = null) {
-    const values = mapValues(schemaDefinition, (obj, name) => ({
+    this.schemaNames = []
+    this.byDoctype = {}
+    this.add(schemaDefinition)
+    this.client = client
+  }
+
+  add(schemaDefinition = {}) {
+    const namesToAdd = Object.keys(schemaDefinition)
+
+    if (intersectWith(namesToAdd, this.schemaNames))
+      throw new Error(
+        `Cannot add schemas with name(s): ${intersection(
+          namesToAdd,
+          this.schemaNames
+        ).join(', ')} (name(s) already exist(s)`
+      )
```

Comment:
I find that we should strive to have the happy case as clear as possible and keep checks very short. I feel that the logic of the method is shadowed by the checks.

```js
const ensureCanBeAdded = (schema, otherSchema) => {
  const sameNames = intersectionBy(newSchemas, existingSchemas, x => x.name)
  assert(sameNames.length === 0, 'Duplicated names in schemas being added: ' + sameNames)

  const sameDoctypes = intersectionBy(newSchemas, existingSchemas, x => x.doctype)
  assert(sameDoctypes.length === 0, 'Duplicated doctypes in schemas being added: ' + sameDoctypes)
}

add(schemaDefinition={}) {
  // Normalize as early as possible to deal with clean objects as soon as possible
  const newSchemas = normalizeSchemaDefinition(schemaDefinition)
  const existingSchemas = Object.values(this.byDoctype)
  ensureCanBeAdded(newSchemas, existingSchemas)
   
  ....
}
```

Reply:
I agree, I am not plenty satisfied by the current implementation. Let's use your suggestion. 👍

Reply:
And good point about the normalization

Reply:
intersectionBy is in lodash but not assert, I add it via `const assert = (cond msg) => { if (!cond) { throw new Error(msg) } }

Reply:
Tell me if the last version suits you (I rebased).